### ISSUE TITLE
Mouse cursor doesn't follow pen position on Wayland

### ISSUE THREAD

**Issue by maximilianum** [OpenBoard-org/OpenBoard#554]
I'm using Fedora 35 with GNOME 41 on Wayland, the mouse cursor is not following the pen position on the tablet making very difficult to draw accurately. If I start the session with Xorg it works correctly.

**Comment by letsfindaway** [comment]
This might be a general problem with Wayland. In Wayland, by design, a graphic tablet has its own pointer and does not "steal" the mouse pointer. However OpenBoard assumes that those are the same. For the small cross used as pointer for drawing OpenBoard changes the mouse pointer shape. But the mouse pointer remains at another position.

Other people have similar problems with other applications (e.g. https://gitlab.gnome.org/GNOME/gimp/-/issues/5334), so it is not specific to OpenBoard, but to the way how Wayland handles tablet and touch pointers.

What you could probably do (besides using Xorg) is to set the threshold for changing the cross to a small circle to "1 px" in the settings:

![grafik](https://user-images.githubusercontent.com/35591413/151038346-b7a4f784-a7ae-4022-867f-d5a2e8a9a7b8.png)

This small circle is actually not a pointer, but drawn by OpenBoard, and will probably be at the right position. I haven't tested it, but this might work.

**Comment by maximilianum** [comment]
> This might be a general problem with Wayland. In Wayland, by design, a graphic tablet has its own pointer and does not "steal" the mouse pointer. However OpenBoard assumes that those are the same. For the small cross used as pointer for drawing OpenBoard changes the mouse pointer shape. But the mouse pointer remains at another position.
> 
> Other people have similar problems with other applications (e.g. https://gitlab.gnome.org/GNOME/gimp/-/issues/5334), so it is not specific to OpenBoard, but to the way how Wayland handles tablet and touch pointers.
> 
> What you could probably do (besides using Xorg) is to set the threshold for changing the cross to a small circle to "1 px" in the settings:
> 
> ![grafik](https://user-images.githubusercontent.com/35591413/151038346-b7a4f784-a7ae-4022-867f-d5a2e8a9a7b8.png)
> 
> This small circle is actually not a pointer, but drawn by OpenBoard, and will probably be at the right position. I haven't tested it, but this might work.

I've just tested setting the threshold to 1 px and it works, I can see the small circle following the pen. I still need to use the mouse in order to select a different tool, but it's not a big problem.

**Comment by maximilianum** [comment]
With Gimp 2.10 it's true the mouse pointer doesn't follow the pen, but another pointer (a perfect copy of the mouse pointer) is moving with the pen and you can draw, select menus and tools with it.

**Comment by letsfindaway** [comment]
Let me rephrase what I said:

It is not a "problem" of Wayland, it is a new feature allowing touch and tablet devices to have their own pointer, independent of the mouse cursor. In practice, this is a very good idea. I always found it annoying that when using the mouse and tablet simultaneously on different screens, the cursor was jumping to the tablet when I use the pen and I have to move it back when I use the mouse, e.g. on another screen. So: no doubt, good idea.

But on the other hand applications need to be prepared to handle this. And also frameworks need to provide support.

When looking over the Qt API - even Qt 6 - I do not see any API to handle multiple cursors or cursor shapes. QCursor still just refers to the mouse cursor. You can distinguish the devices from the events they generate. This is already possible with Qt5 and this is how OpenBoard handles this. But there is no way to affect the tablet cursor shape and set it e.g. to the small cross - at least I did not find a way.

So before OpenBoard can really solve this, an API must be available. In the meantime it can only use a workaround by drawing the cursor by itself using the tablet events. This is how the small circle is drawn.

Of course there are interfaces which offer the necessary functionality and gimp - at least actual versions - are probably using them. But I still wonder how a pure Qt application like OpenBoard can and should handle this in a platform independent way. Remember: OpenBoard must offer the same functionality for MacOS and Windows, too and relies on Qt to provide this platform independence.
